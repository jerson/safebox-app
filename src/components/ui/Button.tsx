import React from 'react';
import {
  GestureResponderEvent,
  StyleProp,
  StyleSheet,
  TextStyle,
  View,
  ViewStyle
} from 'react-native';
import Touchable, { TouchableProps } from './Touchable';
import Colors from '../../modules/constants/Colors';
import Loading from './Loading';
import Text from './Text';
import Icon from 'react-native-vector-icons/Feather';

const styles = StyleSheet.create({
  container: {
    padding: 12,
    borderRadius: 10,
    margin: 0,
    zIndex: 1,
    flexDirection: 'row',
    justifyContent: 'center'
  } as ViewStyle,
  containerDisabled: {
    backgroundColor: Colors.grey3,
    borderColor: Colors.grey3
  } as ViewStyle,
  loading: {
    marginRight: 4
  } as ViewStyle,
  icon: {
    fontSize: 14,
    top: 3,
    color: Colors.white,
    marginRight: 4
  } as TextStyle,
  title: {
    textAlign: 'center',
    color: Colors.white,
    fontSize: 14
  } as TextStyle,
  containerLoading: {
    position: 'absolute',
    left: 10,
    top: 0,
    bottom: 0,
    justifyContent: 'center'
  } as ViewStyle,
  containerLoadingSmall: {
    left: 4
  } as ViewStyle,
  containerLoadingCenter: {
    position: 'relative',
    left: 0
  } as ViewStyle,
  light: {
    backgroundColor: Colors.white
  } as ViewStyle,
  primaryLight: {
    backgroundColor: Colors.primaryLight,
    borderColor: Colors.primaryLight
  } as ViewStyle,
  primary: {
    backgroundColor: Colors.primary,
    borderColor: Colors.primary
  } as ViewStyle,
  secondary: {
    backgroundColor: Colors.secondary,
    borderColor: Colors.secondary
  } as ViewStyle,
  accent: {
    backgroundColor: Colors.accent,
    borderColor: Colors.accent
  } as ViewStyle,
  accentDark: {
    backgroundColor: Colors.accentDark,
    borderColor: Colors.accentDark
  } as ViewStyle,
  danger: {
    backgroundColor: Colors.danger,
    borderColor: Colors.danger
  } as ViewStyle,
  disabled: {
    backgroundColor: Colors.grey4,
    borderColor: Colors.grey4
  } as ViewStyle,
  small: {
    padding: 4,
    paddingHorizontal: 12
  } as ViewStyle,
  center: {
    minHeight: 30,
    justifyContent: 'center'
  } as ViewStyle
});

export interface ButtonProps extends TouchableProps {
  isLoading?: boolean;
  disabled?: boolean;
  allowOnPress?: boolean;
  /**
   * cuando `title` es undefined el loading se centra en el boton
   */
  title?: string;
  textStyle?: StyleProp<TextStyle>;
  icon?: string;
  iconStyle?: StyleProp<TextStyle>;
  light?: boolean;
  small?: boolean;
  typeColor?:
    | 'primaryLight'
    | 'primary'
    | 'secondary'
    | 'accentDark'
    | 'accent'
    | 'danger';
}

function Button(props: ButtonProps) {
  const {
    style,
    disabled,
    title,
    textStyle,
    icon,
    iconStyle,
    isLoading,
    onPress,
    light,
    typeColor,
    small,
    allowOnPress,
    ...extraProps
  } = props;

  const onPressCallback = (event: GestureResponderEvent) => {
    if ((disabled && !allowOnPress) || isLoading) {
      return;
    }
    if (typeof onPress === 'function') {
      onPress(event);
    }
  };

  const stylesButton = [];
  let color;

  switch (typeColor) {
    case 'primaryLight':
      stylesButton.push(styles.primaryLight);
      color = Colors.white;
      break;
    case 'primary':
      stylesButton.push(styles.primary);
      color = Colors.white;
      break;
    case 'secondary':
      stylesButton.push(styles.secondary);
      color = Colors.white;
      break;
    case 'accentDark':
      stylesButton.push(styles.accentDark);
      color = Colors.white;
      break;
    case 'accent':
      stylesButton.push(styles.accent);
      color = Colors.white;
      break;
    case 'danger':
      stylesButton.push(styles.danger);
      color = Colors.white;
      break;
    default:
      break;
  }

  if (disabled) {
    stylesButton.push(styles.disabled);
    color = Colors.grey5;
  }
  if (small) {
    stylesButton.push(styles.small);
  }

  return (
    <Touchable
      activeOpacity={disabled ? 1 : 0.2}
      style={[
        styles.container,
        stylesButton,
        disabled && styles.containerDisabled,
        light && styles.light,
        style,
        !title && styles.center
      ]}
      testID={'button_touchable'}
      {...extraProps}
      onPress={onPress ? onPressCallback : undefined}
    >
      {!!icon && !isLoading && (
        <Icon style={[styles.icon, { color }, iconStyle]} name={icon} />
      )}
      {isLoading && (
        <View
          style={[
            styles.containerLoading,
            small && styles.containerLoadingSmall,
            !title && styles.containerLoadingCenter
          ]}
        >
          <Loading style={styles.loading} size={'small'} color={color} />
        </View>
      )}
      {!!title && (
        <Text style={[styles.title, { color }, textStyle]}>{title}</Text>
      )}
    </Touchable>
  );
}

Button.defaultProps = {
  typeColor: 'primary'
};
export default Button;
