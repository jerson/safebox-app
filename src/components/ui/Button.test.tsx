import React from 'react';
import * as renderer from 'react-test-renderer';

import Button, { ButtonProps } from './Button';

const props: ButtonProps = {
  title: 'Sample',
  onPress: jest.fn()
};
test('renders without crashing', () => {
  const snapshot = renderer.create(<Button {...props} />).toJSON();
  expect(snapshot).toBeTruthy();
});

test('render just views', () => {
  const snapshot = renderer.create(<Button {...props} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});

test('isLoading', async () => {
  const customProps: ButtonProps = {
    ...props,
    isLoading: true
  };
  const snapshot = renderer.create(<Button {...customProps} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});

test('disabled', async () => {
  const customProps: ButtonProps = {
    ...props,
    disabled: true
  };
  const snapshot = renderer.create(<Button {...customProps} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});

test('allowOnPress', async () => {
  const customProps: ButtonProps = {
    ...props,
    allowOnPress: false
  };
  const snapshot = renderer.create(<Button {...customProps} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});

test('icon', async () => {
  const customProps: ButtonProps = {
    ...props,
    icon: 'trash'
  };
  const snapshot = renderer.create(<Button {...customProps} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});
test('light', async () => {
  const customProps: ButtonProps = {
    ...props,
    light: true
  };
  const snapshot = renderer.create(<Button {...customProps} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});
test('small', async () => {
  const customProps: ButtonProps = {
    ...props,
    small: true
  };
  const snapshot = renderer.create(<Button {...customProps} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});

test('typeColor:primary', async () => {
  const customProps: ButtonProps = {
    ...props,
    typeColor: 'primary'
  };
  const snapshot = renderer.create(<Button {...customProps} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});
test('typeColor:secondary', async () => {
  const customProps: ButtonProps = {
    ...props,
    typeColor: 'secondary'
  };
  const snapshot = renderer.create(<Button {...customProps} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});
test('typeColor:accentDark', async () => {
  const customProps: ButtonProps = {
    ...props,
    typeColor: 'accentDark'
  };
  const snapshot = renderer.create(<Button {...customProps} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});
test('typeColor:accent', async () => {
  const customProps: ButtonProps = {
    ...props,
    typeColor: 'accent'
  };
  const snapshot = renderer.create(<Button {...customProps} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});

test('typeColor:danger', async () => {
  const customProps: ButtonProps = {
    ...props,
    typeColor: 'accent'
  };
  const snapshot = renderer.create(<Button {...customProps} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});

test('button_touchable press', async () => {
  const instance = renderer.create(<Button {...props} />).root;

  const element = instance.findByProps({ testID: 'button_touchable' });
  element.props.onPress();

  expect(props.onPress).toBeCalled();
});

test('button_touchable press disabled', async () => {
  const customProps: ButtonProps = {
    ...props,
    onPress: undefined,
    disabled: true,
    allowOnPress: false
  };
  const instance = renderer.create(<Button {...customProps} />).root;

  const element = instance.findByProps({ testID: 'button_touchable' });

  expect(typeof element.props.onPress).toBe('undefined');
});
