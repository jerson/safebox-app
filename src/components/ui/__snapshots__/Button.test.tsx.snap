// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allowOnPress 1`] = `
<_class
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#4531cd",
      "borderColor": "#4531cd",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "margin": 0,
      "opacity": 1,
      "padding": 12,
      "zIndex": 1,
    }
  }
  testID="button_touchable"
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Nunito",
          "fontSize": undefined,
          "fontStyle": "normal",
          "fontWeight": "400",
        },
        Object {},
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 14,
            "textAlign": "center",
          },
          undefined,
          undefined,
        ],
      ]
    }
  >
    Sample
  </Text>
</_class>
`;

exports[`disabled 1`] = `
<_class
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#ebf1f6",
      "borderColor": "#ebf1f6",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "margin": 0,
      "opacity": 1,
      "padding": 12,
      "zIndex": 1,
    }
  }
  testID="button_touchable"
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Nunito",
          "fontSize": undefined,
          "fontStyle": "normal",
          "fontWeight": "400",
        },
        Object {},
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 14,
            "textAlign": "center",
          },
          undefined,
          undefined,
        ],
      ]
    }
  >
    Sample
  </Text>
</_class>
`;

exports[`icon 1`] = `
<_class
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#4531cd",
      "borderColor": "#4531cd",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "margin": 0,
      "opacity": 1,
      "padding": 12,
      "zIndex": 1,
    }
  }
  testID="button_touchable"
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 12,
        },
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 14,
            "marginRight": 4,
            "top": 3,
          },
          undefined,
        ],
        Object {
          "fontFamily": "Feather",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    ï‡§
  </Text>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Nunito",
          "fontSize": undefined,
          "fontStyle": "normal",
          "fontWeight": "400",
        },
        Object {},
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 14,
            "textAlign": "center",
          },
          undefined,
          undefined,
        ],
      ]
    }
  >
    Sample
  </Text>
</_class>
`;

exports[`isLoading 1`] = `
<_class
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#4531cd",
      "borderColor": "#4531cd",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "margin": 0,
      "opacity": 1,
      "padding": 12,
      "zIndex": 1,
    }
  }
  testID="button_touchable"
>
  <_class
    style={
      Array [
        Object {
          "bottom": 0,
          "justifyContent": "center",
          "left": 10,
          "position": "absolute",
          "top": 0,
        },
        undefined,
        false,
      ]
    }
  >
    <_class
      style={
        Array [
          false,
          false,
          false,
          Object {
            "marginRight": 4,
          },
        ]
      }
    >
      <ActivityIndicator
        animating={true}
        color="#FFFFFF"
        hidesWhenStopped={true}
        size="small"
      />
    </_class>
  </_class>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Nunito",
          "fontSize": undefined,
          "fontStyle": "normal",
          "fontWeight": "400",
        },
        Object {},
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 14,
            "textAlign": "center",
          },
          undefined,
          undefined,
        ],
      ]
    }
  >
    Sample
  </Text>
</_class>
`;

exports[`light 1`] = `
<_class
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "borderColor": "#4531cd",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "margin": 0,
      "opacity": 1,
      "padding": 12,
      "zIndex": 1,
    }
  }
  testID="button_touchable"
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Nunito",
          "fontSize": undefined,
          "fontStyle": "normal",
          "fontWeight": "400",
        },
        Object {},
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 14,
            "textAlign": "center",
          },
          Object {
            "color": "#4531cd",
          },
          undefined,
        ],
      ]
    }
  >
    Sample
  </Text>
</_class>
`;

exports[`render just views 1`] = `
<_class
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#4531cd",
      "borderColor": "#4531cd",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "margin": 0,
      "opacity": 1,
      "padding": 12,
      "zIndex": 1,
    }
  }
  testID="button_touchable"
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Nunito",
          "fontSize": undefined,
          "fontStyle": "normal",
          "fontWeight": "400",
        },
        Object {},
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 14,
            "textAlign": "center",
          },
          undefined,
          undefined,
        ],
      ]
    }
  >
    Sample
  </Text>
</_class>
`;

exports[`small 1`] = `
<_class
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#4531cd",
      "borderColor": "#4531cd",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "margin": 0,
      "opacity": 1,
      "padding": 4,
      "paddingHorizontal": 12,
      "zIndex": 1,
    }
  }
  testID="button_touchable"
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Nunito",
          "fontSize": undefined,
          "fontStyle": "normal",
          "fontWeight": "400",
        },
        Object {},
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 14,
            "textAlign": "center",
          },
          undefined,
          undefined,
        ],
      ]
    }
  >
    Sample
  </Text>
</_class>
`;

exports[`typeColor:accent 1`] = `
<_class
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#c2d0e7",
      "borderColor": "#c2d0e7",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "margin": 0,
      "opacity": 1,
      "padding": 12,
      "zIndex": 1,
    }
  }
  testID="button_touchable"
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Nunito",
          "fontSize": undefined,
          "fontStyle": "normal",
          "fontWeight": "400",
        },
        Object {},
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 14,
            "textAlign": "center",
          },
          undefined,
          undefined,
        ],
      ]
    }
  >
    Sample
  </Text>
</_class>
`;

exports[`typeColor:accentDark 1`] = `
<_class
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#6781a6",
      "borderColor": "#6781a6",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "margin": 0,
      "opacity": 1,
      "padding": 12,
      "zIndex": 1,
    }
  }
  testID="button_touchable"
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Nunito",
          "fontSize": undefined,
          "fontStyle": "normal",
          "fontWeight": "400",
        },
        Object {},
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 14,
            "textAlign": "center",
          },
          undefined,
          undefined,
        ],
      ]
    }
  >
    Sample
  </Text>
</_class>
`;

exports[`typeColor:danger 1`] = `
<_class
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#c2d0e7",
      "borderColor": "#c2d0e7",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "margin": 0,
      "opacity": 1,
      "padding": 12,
      "zIndex": 1,
    }
  }
  testID="button_touchable"
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Nunito",
          "fontSize": undefined,
          "fontStyle": "normal",
          "fontWeight": "400",
        },
        Object {},
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 14,
            "textAlign": "center",
          },
          undefined,
          undefined,
        ],
      ]
    }
  >
    Sample
  </Text>
</_class>
`;

exports[`typeColor:primary 1`] = `
<_class
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#4531cd",
      "borderColor": "#4531cd",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "margin": 0,
      "opacity": 1,
      "padding": 12,
      "zIndex": 1,
    }
  }
  testID="button_touchable"
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Nunito",
          "fontSize": undefined,
          "fontStyle": "normal",
          "fontWeight": "400",
        },
        Object {},
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 14,
            "textAlign": "center",
          },
          undefined,
          undefined,
        ],
      ]
    }
  >
    Sample
  </Text>
</_class>
`;

exports[`typeColor:secondary 1`] = `
<_class
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#65748b",
      "borderColor": "#65748b",
      "borderRadius": 10,
      "flexDirection": "row",
      "justifyContent": "center",
      "margin": 0,
      "opacity": 1,
      "padding": 12,
      "zIndex": 1,
    }
  }
  testID="button_touchable"
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "Nunito",
          "fontSize": undefined,
          "fontStyle": "normal",
          "fontWeight": "400",
        },
        Object {},
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 14,
            "textAlign": "center",
          },
          undefined,
          undefined,
        ],
      ]
    }
  >
    Sample
  </Text>
</_class>
`;
